### PR TITLE
Stripe Credit Card gateway now uses the Gateway API

### LINKED ISSUE #6181: Stripe - Credit Card
## Acceptance Criteria

- [x] Creates a payment using credit card fields.
- [x] Redirects for 3D Secure when requires action.

### PR DESCRIPTION
**Pull request by kjohnson**
Resolves #6181

Depends on #6179, #6189

## Description

This PR registers a new `stripe` gateway for handling credit card payments. This is the default gateway for Stripe.

The basic change is refactoring the gateway logic to use the new `GatewayPaymentData $paymentData` input and `GatewayCommand` return values.

The implementation of the gateway processing has also been refactored in order to support automated testing.

Other refactors include replacing nested `if/then/else` logic with early returns and custom exceptions (extending `PaymentGatewayException`). This co-locates the expectation of an error with the code block resulting in the error.

## Affects

The legacy implementation of the `stripe` gateway (`class-give-stripe-card.php`) has an early return which prevents the legacy process from running. This allows for the continued use of the legacy code for gateway specific settings.

## Testing Instructions

This replaces the Stripe Credit Card gateway implementation. Enable the Stripe Credit Card Gateway and complete a payment. There should be no change in behavior.

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit and/or end-to-end tests
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

File: src/Framework/PaymentGateways/PaymentGateway.php [hunk 1]
```diff
@@ -146,6 +148,16 @@ public function handleGatewayPaymentCommand(GatewayCommand $command, GatewayPaym
             $this->handleResponse($response);
         }
 
+        if ($command instanceof PaymentProcessing) {
```
Comment (kjohnson) [inline src/Framework/PaymentGateways/PaymentGateway.php:151]: This change has been `cherry-pick`ed into a separate PR, see https://github.com/impress-org/givewp/pull/6179.

File: includes/gateways/stripe/includes/class-give-stripe-gateway.php [hunk 2]
```diff
@@ -24,79 +25,89 @@
...
+
+            /**
+             * @unreleased
+             *
+             * Once a gateway has been moved over to our Payment Gateway API
+             * we do not want the legacy processing to run.
+             */
+            if (!$paymentGatewayRegister->hasPaymentGateway($this->id)) {
+                add_action("give_gateway_{$this->id}", [$this, 'process_payment']);
+            }
```
Comment (jonwaldstein) [inline includes/gateways/stripe/includes/class-give-stripe-gateway.php:109]: @kjohnson I saw you were returning early to prevent legacy processing.  We don't want to get in a situation where both our legacy and gateway api hooks are running so I added this conditional to make our lives easier.  We'll want do this for other gateway migrations as well.

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function getLegacyFormFieldMarkup($formId)
+    {
+        if (FormUtils::isLegacyForm($formId)) {
+            return false;
+        }
+
+        // @TODO Migrate field markup from legacy gateway implementation.
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:142]: @kjohnson what's our plan here?  Is there any reason we can't move this over in this PR?

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        switch( $intent->status() )  {
+            case 'requires_action':
+                give_insert_payment_note( $paymentData->donationId, 'Stripe requires additional action to be fulfilled.' );
+                give_update_meta( $paymentData->donationId, '_give_stripe_payment_intent_require_action_url', $intent->nextActionRedirectUrl() );
+                return new RedirectOffsite( $intent->nextActionRedirectUrl() );
+            case 'succeeded':
+                return new PaymentProcessing( $intent->id() );
+            default:
+                throw new PaymentIntentException( sprintf( __( 'Unhandled payment intent status: %s', 'give' ), $intent->status() ) );
+        }
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:98]: @kjohnson it looks like you got all the logic moved over, awesome!  is there anyway we can make this monster more readable?  It's just a little hard to follow for a simpleton like myself.

File: src/PaymentGateways/Gateways/Stripe/LegacyDonationData.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Give\PaymentGateways\Gateways\Stripe;
+
+use Give\PaymentGateways\DataTransferObjects\GatewayPaymentData;
+
+/**
+ * @unreleased
+ */
+class LegacyDonationData
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/Stripe/LegacyDonationData.php:10]: Let's move these Legacy prefixed classes into the `LegacyPaymentGateways` domain.  That's where our legacy adapters and purgatory items live that kind of bridge the gap into our new api.  Let's try to keep our PaymentGateways domain as clean as possible.

File: src/PaymentGateways/Gateways/Stripe/LegacyStripePaymentIntent.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace Give\PaymentGateways\Gateways\Stripe;
+
+use Give_Stripe_Payment_Intent;
+
+/**
+ * @unreleased
+ */
+class LegacyStripePaymentIntent
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/Stripe/LegacyStripePaymentIntent.php:10]: Another candidate for the `LegacyPaymentGateways` domain.

File: src/PaymentGateways/Gateways/Stripe/Helpers.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Give\PaymentGateways\Gateways\Stripe;
+
+/**
+ * @unreleased
+ */
+class Helpers
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/Stripe/Helpers.php:8]: Another candidate for the `LegacyPaymentGateways` domain.

**Review by jonwaldstein — CHANGES_REQUESTED** [review]
@kjohnson it works! nice job migrating this stuff over.  Just a few suggestions.

File: includes/gateways/stripe/includes/class-give-stripe-gateway.php (hunk 2 above)
Comment (kjohnson) [reply includes/gateways/stripe/includes/class-give-stripe-gateway.php:109]: @jonwaldstein that works too. I was wanting to restrict my changes to the gateway that I was migrating, but this is where it would eventually end up - that works for me 👍🏻

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php (hunk 3 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:142]: @jonwaldstein I was considering doing that in a separate PR, moving a logical piece at a time.

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php (hunk 4 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:98]: @jonwaldstein this is the more readable version 😅. I do agree that more can be done, I just found myself reverting changes a lot, so I haven't settled on a direction yet. My goal in this PR is to get the logic into the Gateway API and then refactor from there since we will have some tests in place.

Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:98]: I'm probably going to introduce some `Action` classes to group blocks of code that are related, which would make this class more `Controller`-like.

Comment (jonwaldstein) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:98]: @kjohnson sounds good, actions or local methods to group blocks of code would be great - thanks!

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php (hunk 3 above)
Comment (jonwaldstein) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php:142]: I would encourage moving that over now if you're doing one PR for each Stripe gateway variant so that it covers the majority of the gateway.  All depends on your workflow of course.

File: src/PaymentGateways/Gateways/Stripe/LegacyDonationData.php (hunk 5 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/LegacyDonationData.php]: @jonwaldstein I was _almost_ able to refactor out the need for the `LegacyDonationData` class, but alas the format is needed for backwards compatibility with the filter used when preparing Stripe metadata.

Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/LegacyDonationData.php]: That said, I have relocated the `LegacyDonationData` to `src/LegacyPaymentGateways/Adapters`.

File: src/PaymentGateways/Gateways/Stripe/Helpers.php (hunk 7 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Helpers.php]: @jonwaldstein this helper class was refactored out - no longer necessary.

File: src/PaymentGateways/Gateways/Stripe/LegacyStripePaymentIntent.php (hunk 6 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/LegacyStripePaymentIntent.php]: Refactored to be a payment intent value object.

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php (hunk 4 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php]: @jonwaldstein big refactor that moves the procedural processing to `Action` classes.

**Comment by kjohnson** [comment]
@jonwaldstein I just pushed a big refactor the encapsulate processing logic into `Action` classes.

This refactor also cleans up how the different dependencies are juggled between the encapsulated steps.

I found it convoluted to manage the dependencies passed form action to action to action, so I introduced a "shared context" between the steps. my solution is a combination of the pipeline and dependency injection patterns. 

The resulting code should be easy to understand at a glance and the encapsulated processes should be easy to maintain.

```php
public function createPayment(GatewayPaymentData $paymentData)
{
    $workflow = new Workflow();
    $workflow->bind( $paymentData );

    $workflow->action(new Actions\GetPaymentMethodFromRequest);
    $workflow->action(new Actions\SaveDonationSummary);
    $workflow->action(new Actions\GetOrCreateStripeCustomer);
    $workflow->action(new Actions\CreatePaymentIntent);

    return $this->handlePaymentIntentStatus(
        $workflow->resolve( PaymentIntent::class )
    );
}
```

File: src/Framework/PaymentGateways/PaymentGateway.php (hunk 1 above)
Comment (ravinderk) [reply src/Framework/PaymentGateways/PaymentGateway.php:151]: PR has been merged.

File: includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php [hunk 8]
```diff
@@ -147,6 +147,12 @@ public function addCreditCardForm( $form_id, $args, $echo = true ) {
 		 */
 		public function process_payment( $donation_data ) {
 
+            /**
+             * Early return to prevent legacy processing of the Stripe Credit Card gateway.
+             * See src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php
+             */
+            return;
```
Comment (ravinderk) [inline includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php:154]: @kjohnson I think it is not a request because you already handle this in the main class `Give_Stripe_Gateway`.  This means legacy payment processing logic will not run if the payment method is registered with a new gateway API.

Can we remove this?

File: src/LegacyPaymentGateways/Adapters/LegacyDonationData.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Give\LegacyPaymentGateways\Adapters;
+
+use Give\PaymentGateways\DataTransferObjects\GatewayPaymentData;
+
+/**
+ * @unreleased
+ */
+class LegacyDonationData
```
Comment (ravinderk) [inline src/LegacyPaymentGateways/Adapters/LegacyDonationData.php:10]: @kjohnson Maybe we do not require this. I find out that you are using this for function `give_stripe_prepare_metadata`. Can we use this opportunity to deprecate and create a new one that will use `GatewayPaymentData`?
If you do not want to refactor code then for this reason I added `legacyDonationData` in `FormData` which is only for internal use. I am coping the same property to create `GatewayPaymentData` which I am using to gracefully deprecate filter and action hook.

Let me know your thoughts.

File: src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+use Give\PaymentGateways\Gateways\Stripe\ValueObjects\PaymentMethod;
+use Give\PaymentGateways\Gateways\Stripe\WorkflowAction;
+use Give\ValueObjects\Money;
+
+class CreatePaymentIntent extends WorkflowAction
+{
+    public function __invoke(
+        GatewayPaymentData $paymentData,
+        DonationSummary $donationSummary,
+        \Give_Stripe_Customer $giveStripeCustomer,
```
Comment (ravinderk) [inline src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php:18]: @kjohnson I agree that you about less rewrite code for gateway migration. Any specific reason to use a legacy class? Can we handle it differently in which we do not require legacy classes?

File: src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        );
+
+        // Send Stripe Receipt emails when enabled.
+        if ( give_is_setting_enabled( give_get_option( 'stripe_receipt_emails' ) ) ) {
+            $intent_args['receipt_email'] = $paymentData->donorInfo->email;
+        }
+
+        $intent = give( PaymentIntent::class )->create( $intent_args );
+
+        give_insert_payment_note( $paymentData->donationId, 'Stripe Charge/Payment Intent ID: ' . $intent->id() );
```
Comment (ravinderk) [inline src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php:50]: @kjohnson do you think we can internationalize these strings?

File: src/PaymentGateways/Gateways/Stripe/ValueObjects/DonationSummary.php [hunk 12]
```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Give\PaymentGateways\Gateways\Stripe\ValueObjects;
+
+use Give\PaymentGateways\DataTransferObjects\GatewayPaymentData;
+
+class DonationSummary
```
Comment (ravinderk) [inline src/PaymentGateways/Gateways/Stripe/ValueObjects/DonationSummary.php:7]: @kjohnson  In most payment gateway, we send donation summary for gateway payment. I can see that this will be useful if we move this class to `framework/Paymentgateways`.

What do you think?

File: tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+use Give\PaymentGateways\DataTransferObjects\GatewayPaymentData;
+use Give\PaymentGateways\Gateways\Stripe\CreditCardGateway;
+
+/**
+ * @unreleased
+ */
+class CreditCardGatewayTest extends Give_Unit_Test_Case
+{
+    /** @test */
+    public function it_creates_a_payment_that_is_processing()
```
Comment (ravinderk) [inline tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php:14]: @kjohnson can we follow the [camel case ](https://en.wikipedia.org/wiki/Camel_case)of the function name to main train consistency?

File: tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        $_POST['give_stripe_payment_method'] = 'pm_1234';
```
Comment (ravinderk) [inline tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php:16]: @kjohnson I was wondering about using global constants ( $_POST, $_GET) in tests but I find out that the `WP_UnitTestCase` class reset global constants after each test.

👍

File: includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php (hunk 8 above)
Comment (kjohnson) [reply includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php:154]: Yeah, I think Jon made a related change elsewhere, but this hadn't been removed.

File: src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php (hunk 10 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php:18]: There is a lot that I _could_ refactor, which I avoided in this PR in order to limit the scope. The goal here is to update the gateway to use the new Gateway API. A part of that is making new code testable, so I made some additional refactors to support the primary goal. Some aspects were skipped for sake of capacity.

In this case, I was able to mock `Give_Stripe_Customer`, so I left it as is.

File: src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php (hunk 11 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php:50]: Sure. A lot of these weren't previously translatable and were brought over via copy/paste. I can update to add translation.

File: src/PaymentGateways/Gateways/Stripe/ValueObjects/DonationSummary.php (hunk 12 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/ValueObjects/DonationSummary.php:7]: My expectation is that the `DonationSummary` gets extracted into a shared Gateway API component. That said, I have chosen to introduce it as a part of the Stripe gateway as a first iterative step.

File: tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php (hunk 13 above)
Comment (kjohnson) [reply tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php:14]: I'm happy to make changes to match the style guide. I would like to discuss [snake_case](https://en.wikipedia.org/wiki/Snake_case) for readability in tests - the methods aren't a public API, so I'm less worried about matching camel case of the codebase. I'll admit that this is a carry over from also working in Laravel.

File: src/LegacyPaymentGateways/Adapters/LegacyDonationData.php (hunk 9 above)
Comment (kjohnson) [reply src/LegacyPaymentGateways/Adapters/LegacyDonationData.php:10]: I think this is outside of the current scope. The goal here is to update to use the new Gateway API. I'm less concerned with the internals of the components used by the gateway - at this time - except where it prevents testing the new code.

File: src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php (hunk 11 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Actions/CreatePaymentIntent.php:50]: @ravinderk resolved by 6d29d4331.

File: includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php (hunk 8 above)
Comment (kjohnson) [reply includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php]: @ravinderk resolved by 19ec5f6.

File: src/PaymentGateways/Gateways/Stripe/Workflow.php [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+            $action->attachWorkflow( $this );
+        }
+
+        $reflection = new ReflectionMethod($action, '__invoke');
+        $action( ...array_map( function( $parameter ) {
+            return $this->resolve( $parameter->getType()->getName() );
+        }, $reflection->getParameters() ) );
+
+        return $this;
+    }
```
Comment (jonwaldstein) [inline src/PaymentGateways/Gateways/Stripe/Workflow.php:43]: @kjohnson sorry bud

Screen Shot 2022-01-28 at 10 45 47 AM

Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Workflow.php:43]: Hm... I probably just need to access it differently. We are using a container in Give, so I'll look there and update.

Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Workflow.php:43]: Found it. I'll update my implementation. Thanks for catching this!

```php
        // Use ReflectionParameter::getClass() prior to its replacement in PHP 7.1
        if (version_compare(PHP_VERSION, '7.1', '<')) {
            $class = $parameter->getClass();

            return $class ? $class->name : null;
        }

        return $parameter->hasType() ? $parameter->getType()->getName() : null;
```

File: includes/gateways/stripe/includes/class-give-stripe-gateway.php (hunk 2 above)
Comment (kjohnson) [reply includes/gateways/stripe/includes/class-give-stripe-gateway.php]: @jonwaldstein I ended up removing the legacy stripe credit card gateway all together, now that I've migrated the legacy form field markup.

File: src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php (hunk 3 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/CreditCardGateway.php]: Resolved by 9650161.

File: src/PaymentGateways/Gateways/Stripe/Workflow.php (hunk 15 above)
Comment (kjohnson) [reply src/PaymentGateways/Gateways/Stripe/Workflow.php]: @jonwaldstein Resolved by 9976f98.

File: tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php (hunk 13 above)
Comment (kjohnson) [reply tests/unit/tests/PaymentGateways/Gateways/Stripe/CreditCardGatewayTest.php:14]: A big difference being that test methods are sentences, while regular methods are names.

**Comment by jonwaldstein** [comment]
@kjohnson i'm getting an error when trying to donate

```
[28-Jan-2022 17:23:34 UTC] PHP Fatal error:  Uncaught Error: Object of class Give\PaymentGateways\Gateways\Stripe\ValueObjects\DonationSummary could not be converted to string in /Users/jonwaldstein/Sites/givewp/wp-content/plugins/givewp/vendor/stripe/stripe-php/lib/Util/Util.php:233
Stack trace:
#0 /Users/jonwaldstein/Sites/givewp/wp-content/plugins/givewp/vendor/stripe/stripe-php/lib/Util/Util.php(172): Stripe\Util\Util::urlEncode(Object(Give\PaymentGateways\Gateways\Stripe\ValueObjects\DonationSummary))
#1 /Users/jonwaldstein/Sites/givewp/wp-content/plugins/givewp/vendor/stripe/stripe-php/lib/HttpClient/CurlClient.php(225): Stripe\Util\Util::encodeParameters(Array)
#2 /Users/jonwaldstein/Sites/givewp/wp-content/plugins/givewp/vendor/stripe/stripe-php/lib/HttpClient/CurlClient.php(283): Stripe\HttpClient\CurlClient->constructRequest('post', 'https://api.str...', Array, Array, false)
#3 /Users/jonwaldstein/Sites/givewp/wp-content/plugins/givewp/vendor/stripe/stripe-php/lib/ApiRequestor.php(455): Stripe\HttpClient\CurlClient->request('post', 'https://ap in /Users/jonwaldstein/Sites/givewp/wp-content/plugins/givewp/vendor/stripe/stripe-php/lib/Util/Util.php on line 233
```

**Comment by kjohnson** [comment]
@jonwaldstein Whoops. I missed that one. The `DonationSummary` originally had a `__toString()` magic method, but the testing doesn't actually create the Stripe Payment Intent so the update was overlooked. Resolved by 116c37c.

**Review by jonwaldstein — APPROVED** [review]
@kjohnson this gateway appears to be working as expected & looks so much cleaner, great job!
